mmtk/src/lib.rs

```diff
@@ -94,30 +93,25 @@ lazy_static! {
 
 type ProcessSlotFn = *const extern "C" fn(closure: Address, slot: Address);
 
-#[repr(C)]
-pub struct Julia_Upcalls {
-    pub scan_julia_exc_obj:
-        extern "C" fn(obj: Address, closure: Address, process_slot: ProcessSlotFn),
-    pub get_stackbase: extern "C" fn(tid: u16) -> usize,
-    pub jl_throw_out_of_memory_error: extern "C" fn(),
-    pub jl_get_gc_disable_counter: extern "C" fn() -> u32,
-    pub mmtk_sweep_malloced_array: extern "C" fn(),
-    pub mmtk_sweep_stack_pools: extern "C" fn(),
-    pub wait_in_a_safepoint: extern "C" fn(),
-    pub exit_from_safepoint: extern "C" fn(old_state: i8),
-    pub jl_hrtime: extern "C" fn() -> u64,
-    pub update_gc_stats: extern "C" fn(u64, usize, bool),
-    pub get_abi_structs_checksum_c: extern "C" fn() -> usize,
-    pub get_thread_finalizer_list: extern "C" fn(tls: OpaquePointer) -> Address,
-    pub get_to_finalize_list: extern "C" fn() -> Address,
-    pub get_marked_finalizers_list: extern "C" fn() -> Address,
-    pub arraylist_grow: extern "C" fn(Address, usize),
-    pub get_jl_gc_have_pending_finalizers: extern "C" fn() -> *mut i32,
-    pub scan_vm_specific_roots: extern "C" fn(closure: *mut crate::slots::RootsWorkClosure),
-    pub update_inlined_array: extern "C" fn(to: Address, from: Address),
-    pub prepare_to_collect: extern "C" fn(),
-    pub get_owner_address: extern "C" fn(m: Address) -> Address,
-    pub mmtk_genericmemory_how: extern "C" fn(m: Address) -> usize,
+#[allow(improper_ctypes)]
+extern "C" {
+    pub fn jl_gc_scan_julia_exc_obj(obj: Address, closure: Address, process_slot: ProcessSlotFn);
```

Comment:
The prefix is a bit confusing to me. What rules did you use for the functions? E.g. the difference between `jl_gc_mmtk_*`, `jl_gc_*` and `jl_*`.

Reply:
To be completely honest, I'm not sure. I've added `jl_gc_*` to most function to emphasise that we're calling them from `gc-mmtk.c` or `gc-common.c`. But now I realised that I missed `jl_hrtime`. The `_mmtk*` part I just left from the previous version, but I can remove it. If you think of a better rule, I'm open to renaming them accordingly.

Reply:
On a second thought, I might have `jl_*` added to the old `UPCALLS` functions we had, similarly to how we have `mmtk_*` in the functions we have in the binding API.

Reply:
We don't need to change the names if the functions are from Julia, or the GC interface. Those are mostly `jl_*` and `jl_gc_*`. 

There are some in `gc-mmtk.c`, and those functions are not part of the GC interface and specific for MMTk -- I think we can have a prefix for those functions, such as `jl_mmtk_*`. 

MMTk functions (exposed from Rust) are still `mmtk_*`.

Hopefully this may make things cleaner. What do you think?

Reply:
Sounds good, I think that makes sense. I'll try to group them a bit better in their declaration in `lib.rs` file as well following this structure (maybe even adding a comment to describe what you mentioned above).